### ISSUE TITLE
Invalid private member diagnostic of template using with decltype of private member method

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#41693]
|  |  |
| --- | --- |
| Bugzilla Link | [42348](https://llvm.org/bz42348) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@smilingthax,@zygoloid |

## Extended Description 
The following code fails to compile with an erroneous error "'Impl' is a private member of 'S<void>'":

```cpp
  template <typename> struct S
    {
    private:
    template <typename> static constexpr auto Impl();

    public:
    template <typename X> using U = decltype(Impl<X>());
    };

  template <typename T> template <typename>
  constexpr auto S<T>::Impl() { }

  using X = S<void>::U<void>;
```

The above code compiles in gcc et al.  Making the following change allows compilation:

```cpp
  template <typename> struct S
    {
    private:
    struct W
      {
      template <typename> static constexpr auto Impl();
      };

    public:
    template <typename X> using U = decltype(W::template Impl<X>());
    };

  template <typename T> template <typename>
  constexpr auto S<T>::W::Impl()
    { }
```

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [42348](https://llvm.org/bz42348) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@smilingthax,@zygoloid |

## Extended Description 
The following code fails to compile with an erroneous error "'Impl' is a private member of 'S&lt;void&gt;'":

```cpp
  template &lt;typename&gt; struct S
    {
    private:
    template &lt;typename&gt; static constexpr auto Impl();

    public:
    template &lt;typename X&gt; using U = decltype(Impl&lt;X&gt;());
    };

  template &lt;typename T&gt; template &lt;typename&gt;
  constexpr auto S&lt;T&gt;::Impl() { }

  using X = S&lt;void&gt;::U&lt;void&gt;;
```

The above code compiles in gcc et al.  Making the following change allows compilation:

```cpp
  template &lt;typename&gt; struct S
    {
    private:
    struct W
      {
      template &lt;typename&gt; static constexpr auto Impl();
      };

    public:
    template &lt;typename X&gt; using U = decltype(W::template Impl&lt;X&gt;());
    };

  template &lt;typename T&gt; template &lt;typename&gt;
  constexpr auto S&lt;T&gt;::W::Impl()
    { }
```
</details>

**Comment by Fznamznon** [comment]
Patch was reverted so reopening this one.

**Comment by EugeneZelenko** [comment]
Fix was done in private branch.

**Comment by Fznamznon** [comment]
Ok, I've made two attempts to fix but both ended up insufficient. The idea was simple - since access checks depend on context:

https://github.com/llvm/llvm-project/blob/2103de0c4cced5d06165a8a8a84cec89c8ab3be0/clang/lib/Sema/SemaAccess.cpp#L1483

, let's start instantiation of a type alias from the context where the alias is declared. This fixes the problem from this issue. However, there is also this piece of code starting from:

https://github.com/llvm/llvm-project/blob/f32662a40b0cc25f779ed10ea6515ba798922df8/clang/lib/Sema/SemaTemplate.cpp#L4054

that checks if some type referenced ATM with a template-id matches current context (or outer context) and this context is a record, then canonical type of the type that is being checked is assigned with `InjectedClassNameType` of the found record. And it seems there might be quite a lot of uncovered by check-clang tests cases where switch of the `CurContext` breaks this whole scheme. I worked around at least two cases in https://github.com/llvm/llvm-project/pull/75069, and received two other reported. So, perhaps there should be another, less obvious approach to fix this.

cc @zygoloid , @cor3ntin , @erichkeane in case you have other ideas.

**Comment by EugeneZelenko** [comment]
Not merged yet.

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-c-11

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [42348](https://llvm.org/bz42348) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@smilingthax,@zygoloid |

## Extended Description 
The following code fails to compile with an erroneous error "'Impl' is a private member of 'S&lt;void&gt;'":

```cpp
  template &lt;typename&gt; struct S
    {
    private:
    template &lt;typename&gt; static constexpr auto Impl();

    public:
    template &lt;typename X&gt; using U = decltype(Impl&lt;X&gt;());
    };

  template &lt;typename T&gt; template &lt;typename&gt;
  constexpr auto S&lt;T&gt;::Impl() { }

  using X = S&lt;void&gt;::U&lt;void&gt;;
```

The above code compiles in gcc et al.  Making the following change allows compilation:

```cpp
  template &lt;typename&gt; struct S
    {
    private:
    struct W
      {
      template &lt;typename&gt; static constexpr auto Impl();
      };

    public:
    template &lt;typename X&gt; using U = decltype(W::template Impl&lt;X&gt;());
    };

  template &lt;typename T&gt; template &lt;typename&gt;
  constexpr auto S&lt;T&gt;::W::Impl()
    { }
```
</details>
